src/SFML/Graphics/Text.cpp

```diff
@@ -280,14 +376,28 @@ std::uint32_t Text::getStyle() const
 ////////////////////////////////////////////////////////////
 const Color& Text::getFillColor() const
 {
-    return m_fillColor;
+    return getCharacterFillColor(0);
+}
+
+
+////////////////////////////////////////////////////////////
+const Color& Text::getCharacterFillColor(unsigned int index) const
+{
+    return m_fillColors[index];
```

Comment:
> What I don't like is that setting color at index N implicitly assigns colors from 0 to N-1. In other words, it fills the left-hand-side range, but not the right.

I don't think that is how it works. When you switch to the multi-color mode, there is this check which makes sure that the vector is at least the string size (in your case, 10).

```cpp
    // Set the new per character colors to the color set by setFillColor()
    if (!m_multiFillColor)
    {
        m_fillColors.resize(m_string.getSize());
        for (std::size_t i = 0; i < m_fillColors.size(); i++)
            m_fillColors[i] = m_fillColors[0];
    }```

If you however provide an index that is out of bounds (ex. 12), the vector will resize to size 13, so that it can keep your index in the array. This is meant for use such as this:
```cpp
Text text;
//add position, rotation, scale, font, style...
//at this point text.setString() has not been called
text.setCharacterFillColor(3, Color::Red);
//this index is 3 which is now out of bounds
//the vector will resize to size 4, so that in case that you set text.setString() later (and the string will have 4 or more characters), 
//it will be seen when you draw it on the screen)
text.setString("Test");
//now the string has been set to "Test" and the vector is still at size 4.
//however the last character will now be red
```



> // now the last character is suddenly blue, 
> // because of stale state from a previous assignment!

As I said earlier, this is completely intended. Since if you create a sf::Text, first set the character to some color before setting the string, the idea was for it to work the same if you first set the string and then set the character color.

> m_perCharacterFillColor? more verbose but clearer.
Idk, to me `m_multiFillColor` is also ok (maybe instead of multi -> multiple)

> Could you add this in a comment?
Sure

> e.g. in loops, you also break when size is reached, not when size-1 is exceeded
Well it seems like some of us think of this differently :)

> This causes UB if the index is out-of-bounds.
Yup, as it says, if you pass an invalid index, you'll get undefined behaviour (though, I should probably mention "out of bounds" somewhere). 
```cpp
    /// The validity of the index is not checked, passing an invalid
    /// index results in an undefined behaviour.```

Reply:
> Yup, as it says, if you pass an invalid index, you'll get undefined behaviour

There's two problems with that:
* First, it's not clear what constitutes "invalid". Anything _less_ than the maximum set index will return a color. Anything _greater_ will cause UB. The user faces a left/right asymmetry for no apparent reason.
* Undefined behavior is an unnecessary source of bugs in applications. It's really just C and C++ where UB is even a socially accepted thing that you just have to live with, any other language tries to prevent it as hard as possible. We should at least have an `assert` here. Some stdlib implementations likely do that inside `operator[]`, but it's not something we can rely on.

Reply:
Hm, one idea of how to fix this is instead of using the vector for both single and multi color modes, keep a sf::Color for single color and a std::vector for multi color. Then if you request an index that is out of bounds, it will return the color stored in single color mode. For example:
```cpp
        text.setString("Test"); //set a string of size 4
        text.setFillColor(Color::Green); //set all characters to green
        text.setCharacterFillColor(3, Color::Blue); //set the fourth character to blue
        Color x = text.getCharacterFillColor(5); //5 is out of bounds for the m_fillColors, therefore, return the green color
```